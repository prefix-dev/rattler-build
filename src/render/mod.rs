#![allow(missing_docs)]
//! Render the dependencies to a final recipe

pub mod pin;
pub mod resolved_dependencies;
mod run_exports;
pub mod solver;
