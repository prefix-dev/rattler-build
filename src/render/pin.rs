//! Evaluate pin expressions and apply them to resolved dependencies
use std::{
    cmp,
    fmt::{Display, Formatter},
    str::FromStr,
};

use rattler_conda_types::{
    MatchSpec, PackageName, ParseStrictness, Version, VersionBumpError, VersionBumpType,
};
use serde::{de, Deserialize, Deserializer, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PinExpression(#[serde(deserialize_with = "deserialize_pin_expression")] String);

fn deserialize_pin_expression<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: Deserializer<'de>,
{
    // A pin expression can only contain x and . (e.g. x.x.x)
    let s = String::deserialize(deserializer)?;
    let s = PinExpression::from_str(&s).map_err(de::Error::custom)?.0;
    Ok(s)
}

impl FromStr for PinExpression {
    type Err = std::io::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.chars().any(|c| c != 'x' && c != '.') {
            return Err(std::io::Error::new(
                std::io::ErrorKind::InvalidInput,
                "Invalid pin expression (can only contain x and .)",
            ));
        }
        Ok(PinExpression(s.to_string()))
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum PinBound {
    Expression(PinExpression),
    Version(Version),
}

impl FromStr for PinBound {
    type Err = std::io::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.chars().any(|c| c != 'x' && c != '.') {
            Ok(PinBound::Version(s.parse().map_err(|e| {
                std::io::Error::new(std::io::ErrorKind::InvalidInput, e)
            })?))
        } else {
            Ok(PinBound::Expression(s.parse()?))
        }
    }
}

impl Display for PinExpression {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

/// A pin to a specific version of a package
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Pin {
    /// The name of the package to pin
    pub name: PackageName,

    /// The pin arguments
    #[serde(flatten)]
    pub args: PinArgs,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PinArgs {
    /// A minimum pin to a version, using `x.x.x...` as syntax
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub lower_bound: Option<PinBound>,

    /// A pin to a version, using `x.x.x...` as syntax
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub upper_bound: Option<PinBound>,

    /// If an exact pin is given, we pin the exact version & hash
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub exact: bool,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub build: Option<String>,
}

impl Default for PinArgs {
    fn default() -> Self {
        Self {
            lower_bound: Some("x.x.x.x.x.x".parse().unwrap()),
            upper_bound: Some("x".parse().unwrap()),
            exact: false,
            build: None,
        }
    }
}

#[derive(Debug, thiserror::Error)]
pub enum PinError {
    #[error("Could not create MatchSpec Pin: {0}")]
    MatchSpec(#[from] std::io::Error),

    #[error("Could not parse version for pinning (element not a number?): {0}")]
    CouldNotPin(String),

    #[error("lower_bound or upper_bound expression is empty string. Needs to be at least `x`")]
    EmptyPinExpression,

    #[error("Could not increment version: {0}")]
    VersionBump(#[from] VersionBumpError),

    #[error("Build specifier and exact=True are not supported together")]
    BuildSpecifierWithExact,
}

pub fn increment(version: &Version, segments: i32) -> Result<Version, VersionBumpError> {
    if segments == 0 {
        return Err(VersionBumpError::InvalidSegment { index: 0 });
    }

    let version = version
        .clone()
        .with_segments(..cmp::min(version.segment_count(), segments as usize))
        .unwrap();

    Ok(version
        .bump(VersionBumpType::Segment(segments - 1))?
        .with_alpha()
        .remove_local()
        .into_owned())
}

impl Pin {
    /// Apply the pin to a version and hash of a resolved package. If a max_pin, min_pin or exact pin
    /// are given, the pin is applied to the version accordingly.
    pub fn apply(&self, version: &Version, build_string: &str) -> Result<MatchSpec, PinError> {
        if self.args.build.is_some() && self.args.exact {
            return Err(PinError::BuildSpecifierWithExact);
        }

        if self.args.exact {
            return Ok(MatchSpec::from_str(
                &format!("{} {} {}", self.name.as_normalized(), version, build_string),
                ParseStrictness::Strict,
            )
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidInput, e))?);
        }

        let mut pin_str = String::new();

        // extract same amount of digits as the pin expression (in the form of x.x.x) from version str
        match self.args.lower_bound.as_ref() {
            Some(PinBound::Expression(expression)) => {
                // number of digits in pin expression
                let pin_digits = expression.0.chars().filter(|c| *c == 'x').count();
                if pin_digits == 0 {
                    return Err(PinError::EmptyPinExpression);
                }

                // get version string up the to pin_digits dot
                let pin = version
                    .clone()
                    .with_segments(..cmp::min(pin_digits, version.segment_count()))
                    .ok_or_else(|| {
                        PinError::CouldNotPin("Failed to extract min_pin from version".to_string())
                    })?;
                pin_str.push_str(&format!(">={}", pin));
            }
            Some(PinBound::Version(version)) => {
                pin_str.push_str(&format!(">={}", version));
            }
            None => {}
        };

        match self.args.upper_bound.as_ref() {
            Some(PinBound::Expression(expression)) => {
                // number of digits in pin expression
                let pin_digits = expression.0.chars().filter(|c| *c == 'x').count();
                if pin_digits == 0 {
                    return Err(PinError::EmptyPinExpression);
                }

                // increment last digit
                let pin = increment(version, pin_digits as i32)?;

                if !pin_str.is_empty() {
                    pin_str.push(',')
                }
                pin_str.push_str(&format!("<{}", pin));
            }
            Some(PinBound::Version(version)) => {
                if !pin_str.is_empty() {
                    pin_str.push(',')
                }
                pin_str.push_str(&format!("<{}", version));
            }
            None => {}
        }

        let name = self.name.as_normalized().to_string();
        let build = self
            .args
            .build
            .as_ref()
            .map(|b| format!(" {}", b))
            .unwrap_or_default();
        Ok(MatchSpec::from_str(
            format!("{name} {pin_str}{build}").as_str().trim(),
            ParseStrictness::Strict,
        )
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidInput, e))?)
    }
}

#[cfg(test)]
mod test {
    use std::path::Path;

    use itertools::Itertools;

    use super::*;

    #[derive(Deserialize)]
    struct TestPinExpression {
        pin: Pin,
        spec: String,
        expected: String,
    }

    #[test]
    fn test_pins() {
        // load data from the test folder
        let p = Path::new(env!("CARGO_MANIFEST_DIR")).join("test-data/pins.yaml");
        let file = fs_err::File::open(p).unwrap();
        let pins: Vec<TestPinExpression> = serde_yaml::from_reader(file).unwrap();

        for test in pins {
            let spec = test.spec;
            // split the spec in 3 parts (name, version, build string)
            let (version, hash) = spec.split_whitespace().collect_tuple().unwrap();
            let version: Version = version.parse().unwrap();
            let spec = test.pin.apply(&version, hash).unwrap();
            println!("{} -> {}", spec, test.expected);
            assert_eq!(spec.to_string(), test.expected);
        }
    }

    #[test]
    fn test_apply_pin() {
        let pin = Pin {
            name: PackageName::from_str("foo").unwrap(),
            args: PinArgs {
                lower_bound: Some("x.x.x".parse().unwrap()),
                upper_bound: Some("x.x.x".parse().unwrap()),
                ..Default::default()
            },
        };

        let version = Version::from_str("1.2.3").unwrap();
        let hash = "1234567890";
        let spec = pin.apply(&version, hash).unwrap();
        assert_eq!(spec.to_string(), "foo >=1.2.3,<1.2.4.0a0");

        let short_version = Version::from_str("1").unwrap();
        let spec = pin.apply(&short_version, hash).unwrap();
        assert_eq!(spec.to_string(), "foo >=1,<1.0.1.0a0");

        let pin = Pin {
            name: PackageName::from_str("foo").unwrap(),
            args: PinArgs {
                upper_bound: Some("x.x.x".parse().unwrap()),
                lower_bound: None,
                ..Default::default()
            },
        };

        let spec = pin.apply(&version, hash).unwrap();
        assert_eq!(spec.to_string(), "foo <1.2.4.0a0");

        let pin = Pin {
            name: PackageName::from_str("foo").unwrap(),
            args: PinArgs {
                lower_bound: Some("x.x.x".parse().unwrap()),
                upper_bound: None,
                ..Default::default()
            },
        };

        let spec = pin.apply(&version, hash).unwrap();
        assert_eq!(spec.to_string(), "foo >=1.2.3");

        let pin = Pin {
            name: "foo".parse().unwrap(),
            args: PinArgs {
                build: Some("foo*".to_string()),
                ..Default::default()
            },
        };

        let spec = pin.apply(&version, hash).unwrap();
        assert_eq!(spec.to_string(), "foo >=1.2.3,<2.0a0 foo*");
    }

    #[test]
    fn test_apply_exact_pin() {
        let pin = Pin {
            name: PackageName::from_str("foo").unwrap(),
            args: PinArgs {
                lower_bound: Some("x.x.x".parse().unwrap()),
                upper_bound: Some("x.x.x".parse().unwrap()),
                exact: true,
                ..Default::default()
            },
        };

        let version = Version::from_str("1.2.3").unwrap();
        let hash = "h1234_0";
        let spec = pin.apply(&version, hash).unwrap();
        assert_eq!(spec.to_string(), "foo ==1.2.3 h1234_0");
    }

    #[test]
    fn test_pin_with_bounds() {
        let pin = Pin {
            name: PackageName::from_str("foo").unwrap(),
            args: PinArgs {
                lower_bound: Some("x.x.x".parse().unwrap()),
                upper_bound: Some("2.4".parse().unwrap()),
                ..Default::default()
            },
        };

        let version = Version::from_str("1.2.3").unwrap();
        let hash = "h1234_0";
        let spec = pin.apply(&version, hash).unwrap();
        assert_eq!(spec.to_string(), "foo >=1.2.3,<2.4");
    }

    #[test]
    fn test_increment() {
        fn increment_to_string(input: &str, segments: i32) -> String {
            let version = Version::from_str(input).unwrap();
            increment(&version, segments).unwrap().to_string()
        }

        assert_eq!(increment_to_string("1.2.3", 3), "1.2.4.0a0");
        assert_eq!(increment_to_string("1.2.3+4.5", 3), "1.2.4.0a0");
        assert_eq!(increment_to_string("1.2.3", 2), "1.3.0a0");
        assert_eq!(increment_to_string("1.2.3", 1), "2.0a0");
        assert_eq!(increment_to_string("1.2.3+3.4", 1), "2.0a0");
        assert_eq!(increment_to_string("9d", 1), "10a");
        assert_eq!(increment_to_string("9d", 2), "9d.1.0a0");
    }
}
