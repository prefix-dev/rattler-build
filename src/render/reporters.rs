use std::{
    sync::{Arc, Mutex},
    time::Duration,
};

use indicatif::{MultiProgress, ProgressBar, ProgressFinish, ProgressStyle};
use rattler::install::Placement;
use rattler_repodata_gateway::{DownloadReporter, JLAPReporter, Reporter};
use url::Url;

/// Reporter used for tracking download progress via `MultiProgress`.
pub struct GatewayReporter {
    progress_bars: Arc<Mutex<Vec<ProgressBar>>>,
    multi_progress: MultiProgress,
    progress_template: Option<ProgressStyle>,
    finish_template: Option<ProgressStyle>,
    prefix: String,
    finish_message: Option<String>,
    steady_tick: Option<Duration>,
    placement: Placement,
}

/// Builder for [`GatewayReporter`].
#[derive(Default)]
pub struct GatewayReporterBuilder {
    multi_progress: Option<MultiProgress>,
    progress_template: Option<ProgressStyle>,
    finish_template: Option<ProgressStyle>,
    prefix: Option<String>,
    finish_message: Option<String>,
    steady_tick: Option<Duration>,
    placement: Option<Placement>,
}

impl GatewayReporter {
    /// Construct a new builder.
    pub fn builder() -> GatewayReporterBuilder {
        GatewayReporterBuilder::default()
    }

    fn place_progress_bar(&self, progress_bar: ProgressBar) -> ProgressBar {
        match &self.placement {
            Placement::Before(other) => self.multi_progress.insert_before(other, progress_bar),
            Placement::After(other) => self.multi_progress.insert_after(other, progress_bar),
            Placement::Index(index) => self.multi_progress.insert(*index, progress_bar),
            Placement::End => self.multi_progress.add(progress_bar),
        }
    }
}

impl DownloadReporter for GatewayReporter {
    fn on_download_start(&self, _url: &Url) -> usize {
        let progress_bar = ProgressBar::new(1)
            .with_finish(ProgressFinish::AndLeave)
            .with_prefix(self.prefix.clone());

        if let Some(template) = &self.progress_template {
            progress_bar.set_style(template.clone());
        }

        if let Some(duration) = self.steady_tick {
            progress_bar.enable_steady_tick(duration);
        }

        let progress_bar = self.place_progress_bar(progress_bar);

        let mut progress_bars = self.progress_bars.lock().unwrap();
        progress_bars.push(progress_bar);
        progress_bars.len() - 1
    }

    fn on_download_complete(&self, _url: &Url, index: usize) {
        if let Some(progress_bar) = self.progress_bars.lock().unwrap().get(index) {
            if let Some(template) = &self.finish_template {
                progress_bar.set_style(template.clone());
            }

            if let Some(message) = &self.finish_message {
                progress_bar.finish_with_message(message.clone());
            } else {
                progress_bar.finish();
            }
        }
    }

    fn on_download_progress(&self, _url: &Url, index: usize, bytes: usize, total: Option<usize>) {
        if let Some(progress_bar) = self.progress_bars.lock().unwrap().get(index) {
            progress_bar.set_length(total.unwrap_or(bytes) as u64);
            progress_bar.set_position(bytes as u64);
        }
    }
}

impl Reporter for GatewayReporter {
    fn jlap_reporter(&self) -> Option<&dyn JLAPReporter> {
        None
    }

    fn download_reporter(&self) -> Option<&dyn DownloadReporter> {
        Some(self)
    }
}

impl GatewayReporterBuilder {
    /// Configure the multi progress instance.
    #[must_use]
    pub fn with_multi_progress(mut self, multi_progress: MultiProgress) -> Self {
        self.multi_progress = Some(multi_progress);
        self
    }

    /// Configure the progress template style.
    #[must_use]
    pub fn with_progress_template(mut self, template: ProgressStyle) -> Self {
        self.progress_template = Some(template);
        self
    }

    /// Configure the finish template style.
    #[must_use]
    pub fn with_finish_template(mut self, template: ProgressStyle) -> Self {
        self.finish_template = Some(template);
        self
    }

    /// Configure the prefix shown for the progress bar.
    #[must_use]
    pub fn with_prefix(mut self, prefix: impl Into<String>) -> Self {
        self.prefix = Some(prefix.into());
        self
    }

    /// Configure the finish message shown when the download completes.
    #[must_use]
    pub fn with_finish_message(mut self, message: impl Into<String>) -> Self {
        self.finish_message = Some(message.into());
        self
    }

    /// Enable steady ticking for the progress bar.
    #[must_use]
    pub fn with_steady_tick(mut self, duration: Duration) -> Self {
        self.steady_tick = Some(duration);
        self
    }

    /// Configure where to place the progress bar alongside other bars.
    #[must_use]
    pub fn with_placement(mut self, placement: Placement) -> Self {
        self.placement = Some(placement);
        self
    }

    /// Finalize the builder.
    pub fn finish(self) -> GatewayReporter {
        GatewayReporter {
            progress_bars: Arc::new(Mutex::new(Vec::new())),
            multi_progress: self.multi_progress.expect("multi progress is required"),
            progress_template: self.progress_template,
            finish_template: self.finish_template,
            prefix: self
                .prefix
                .unwrap_or_else(|| "Downloading repodata".to_string()),
            finish_message: Some(self.finish_message.unwrap_or_else(|| "Done".to_string())),
            steady_tick: self.steady_tick,
            placement: self.placement.unwrap_or_default(),
        }
    }
}
