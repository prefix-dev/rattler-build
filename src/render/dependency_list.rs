use std::{fmt, str::FromStr};

use super::pin::Pin;
use rattler_conda_types::MatchSpec;
use serde::{
    de::{self, Visitor},
    Deserialize, Deserializer, Serialize,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PinSubpackage {
    pub pin_subpackage: Pin,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Compiler {
    pub compiler: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(untagged)]
pub enum Dependency {
    #[serde(deserialize_with = "deserialize_match_spec")]
    Spec(MatchSpec),
    PinSubpackage(PinSubpackage),
    Compiler(Compiler),
}

impl<'de> Deserialize<'de> for Dependency {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct DependencyVisitor;

        impl<'de> Visitor<'de> for DependencyVisitor {
            type Value = Dependency;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str(
                    "a string starting with '__COMPILER', '__PIN_SUBPACKAGE', or a MatchSpec",
                )
            }

            fn visit_str<E>(self, value: &str) -> Result<Dependency, E>
            where
                E: de::Error,
            {
                if let Some(compiler) = value.strip_prefix("__COMPILER ") {
                    Ok(Dependency::Compiler(Compiler {
                        compiler: compiler.to_lowercase(),
                    }))
                } else if let Some(pin) = value.strip_prefix("__PIN_SUBPACKAGE ") {
                    Ok(Dependency::PinSubpackage(PinSubpackage {
                        pin_subpackage: Pin::from_internal_repr(pin),
                    }))
                } else {
                    // Assuming MatchSpec can be constructed from a string.
                    MatchSpec::from_str(value)
                        .map(Dependency::Spec)
                        .map_err(de::Error::custom)
                }
            }
        }

        deserializer.deserialize_str(DependencyVisitor)
    }
}

pub type DependencyList = Vec<Dependency>;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_render_matchspecs() {
        let spec = r#"
- python
- python >=3.6
- python >=3.6,<3.7
- python >=3.6,<3.7.0a0
- python >=3.6,<3.7.0a0[build=py36h6de7cb9_0]
        "#
        .trim();
        let spec: DependencyList = serde_yaml::from_str(spec).unwrap();
        insta::assert_yaml_snapshot!(spec);
    }

    #[test]
    fn test_render_compiler() {
        // The following works, but doesn't play nicely with Jinja (since the jinja statements)
        // are quoted and can't produce proper YAML easily ... so we use the __COMPILER
        // syntax instead. We could / should revisit this!
        // let spec = r#"
        // - { compiler: "c" }
        // - { compiler: "cxx" }
        // - { compiler: "fortran" }
        // - { compiler: "rust" }
        //         "#
        //         .trim();
        //         let spec: DependencyList = serde_yaml::from_str(spec).unwrap();
        //         insta::assert_yaml_snapshot!(spec);
        let spec = r#"
        - __COMPILER C
        - __COMPILER CXX
        - __COMPILER FORTRAN
        - __COMPILER RUST
        "#;
        let spec: DependencyList = serde_yaml::from_str(spec).unwrap();
        insta::assert_yaml_snapshot!(spec);
    }

    #[test]
    fn test_render_pin_subpackage() {
        let pin = "- __PIN_SUBPACKAGE name MAX_PIN= MIN_PIN=x.x.x EXACT=False";
        let spec: DependencyList = serde_yaml::from_str(pin).unwrap();
        insta::assert_yaml_snapshot!(spec);

        let pin = "- __PIN_SUBPACKAGE super-package MAX_PIN=x.x MIN_PIN=x.x.x EXACT=true";
        let spec: DependencyList = serde_yaml::from_str(pin).unwrap();
        let p = &spec[0];
        let p = match p {
            Dependency::PinSubpackage(p) => p,
            _ => panic!("Expected PinSubpackage"),
        };
        assert_eq!(p.pin_subpackage.name.as_normalized(), "super-package");
        assert_eq!(
            p.pin_subpackage.max_pin.as_ref().unwrap().to_string(),
            "x.x"
        );
        assert_eq!(
            p.pin_subpackage.min_pin.as_ref().unwrap().to_string(),
            "x.x.x"
        );
        assert!(p.pin_subpackage.exact);
    }
}
