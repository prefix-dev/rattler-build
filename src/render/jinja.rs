use std::str::FromStr;

use minijinja::{value::Value, Environment, Syntax};
use rattler_conda_types::{Version, VersionSpec};

mod functions {
    use std::str::FromStr;

    use minijinja::Error;

    use crate::render::pin::{Pin, PinExpression};

    pub fn compiler(lang: String) -> Result<String, Error> {
        // we translate the compiler into a YAML string
        Ok(format!("__COMPILER {}", lang.to_lowercase()))
    }

    pub fn pin_subpackage(
        name: String,
        kwargs: Option<minijinja::value::Value>,
    ) -> Result<String, Error> {
        // we translate the compiler into a YAML string
        let mut pin_subpackage = Pin {
            name,
            max_pin: None,
            min_pin: None,
            exact: false,
        };

        let pin_expr_from_value = |pin_expr: &minijinja::value::Value| {
            PinExpression::from_str(&pin_expr.to_string()).map_err(|e| {
                Error::new(
                    minijinja::ErrorKind::SyntaxError,
                    format!("Invalid pin expression: {}", e),
                )
            })
        };

        if let Some(kwargs) = kwargs {
            let max_pin = kwargs.get_attr("max_pin")?;
            if max_pin != minijinja::value::Value::UNDEFINED {
                let pin_expr = pin_expr_from_value(&max_pin)?;
                pin_subpackage.max_pin = Some(pin_expr);
            }
            let min = kwargs.get_attr("min_pin")?;
            if min != minijinja::value::Value::UNDEFINED {
                let pin_expr = pin_expr_from_value(&min)?;
                pin_subpackage.min_pin = Some(pin_expr);
            }
            let exact = kwargs.get_attr("exact")?;
            if exact != minijinja::value::Value::UNDEFINED {
                pin_subpackage.exact = exact.is_true();
            }
        }

        Ok(format!(
            "__PIN_SUBPACKAGE {}",
            pin_subpackage.internal_repr()
        ))
    }
}

/// Get a jinja environment with the correct syntax and functions
pub fn jinja_environment() -> Environment<'static> {
    let mut env = Environment::new();

    env.set_syntax(Syntax {
        block_start: "{%".into(),
        block_end: "%}".into(),
        variable_start: "${{".into(),
        variable_end: "}}".into(),
        comment_start: "#{{".into(),
        comment_end: "}}#".into(),
    })
    .unwrap();

    env.add_function("cmp", |a: &Value, spec: &str| {
        if let Some(version) = a.as_str() {
            // check if version matches spec
            let version = Version::from_str(version).unwrap();
            let version_spec = VersionSpec::from_str(spec).unwrap();
            Ok(version_spec.matches(&version))
        } else {
            // if a is undefined, we are currently searching for all variants and thus return true
            Ok(true)
        }
    });

    env.add_function("compiler", functions::compiler);
    env.add_function("pin_subpackage", functions::pin_subpackage);

    env
}
