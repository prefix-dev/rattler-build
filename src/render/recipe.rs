use core::panic;
use std::collections::BTreeMap;

use minijinja::{self, value::Value, Environment};
use serde_yaml::Value as YamlValue;

use crate::metadata::RenderedRecipe;

/// Given a YAML recipe, this renders all strings it encounters using Jinja
/// templating.
fn render_recipe_recursively(
    recipe: &mut serde_yaml::Mapping,
    jinja_env: &Environment,
    context: &BTreeMap<String, Value>,
) {
    for (_, v) in recipe.iter_mut() {
        match v {
            YamlValue::String(var) => {
                *v = YamlValue::String(jinja_env.render_str(var, context).unwrap());
            }
            YamlValue::Sequence(var) => {
                render_recipe_recursively_seq(var, jinja_env, context);
            }
            YamlValue::Mapping(var) => {
                render_recipe_recursively(var, jinja_env, context);
            }
            _ => {}
        }
    }
}

fn render_recipe_recursively_seq(
    recipe: &mut serde_yaml::Sequence,
    environment: &Environment,
    context: &BTreeMap<String, Value>,
) {
    for v in recipe {
        match v {
            YamlValue::String(var) => {
                *v = YamlValue::String(environment.render_str(var, context).unwrap());
            }
            YamlValue::Sequence(var) => {
                render_recipe_recursively_seq(var, environment, context);
            }
            YamlValue::Mapping(var) => {
                render_recipe_recursively(var, environment, context);
            }
            _ => {}
        }
    }
}

mod functions {
    use std::str::FromStr;

    use minijinja::Error;

    use crate::render::pin::{Pin, PinExpression};

    pub fn compiler(lang: String) -> Result<String, Error> {
        // we translate the compiler into a YAML string
        Ok(format!("__COMPILER {}", lang.to_lowercase()))
    }

    pub fn pin_subpackage(
        name: String,
        kwargs: Option<minijinja::value::Value>,
    ) -> Result<String, Error> {
        // we translate the compiler into a YAML string
        let mut pin_subpackage = Pin {
            name,
            max_pin: None,
            min_pin: None,
            exact: false,
        };

        let pin_expr_from_value = |pin_expr: &minijinja::value::Value| {
            PinExpression::from_str(&pin_expr.to_string()).map_err(|e| {
                Error::new(
                    minijinja::ErrorKind::SyntaxError,
                    format!("Invalid pin expression: {}", e),
                )
            })
        };

        if let Some(kwargs) = kwargs {
            let max_pin = kwargs.get_attr("max_pin")?;
            if max_pin != minijinja::value::Value::UNDEFINED {
                let pin_expr = pin_expr_from_value(&max_pin)?;
                pin_subpackage.max_pin = Some(pin_expr);
            }
            let min = kwargs.get_attr("min_pin")?;
            if min != minijinja::value::Value::UNDEFINED {
                let pin_expr = pin_expr_from_value(&min)?;
                pin_subpackage.min_pin = Some(pin_expr);
            }
            let exact = kwargs.get_attr("exact")?;
            if exact != minijinja::value::Value::UNDEFINED {
                pin_subpackage.exact = exact.is_true();
            }
        }

        Ok(format!(
            "__PIN_SUBPACKAGE {}",
            pin_subpackage.internal_repr()
        ))
    }
}

/// This iteratively renderes the "context" object of a recipe
/// Later values can reference earlier values in the context, for example
///
/// ```yaml
/// context:
///   version: "3.0"
///   version_min: "min_{{ version }}"
/// ```
fn render_context(yaml_context: &serde_yaml::Mapping) -> BTreeMap<String, Value> {
    let mut context = BTreeMap::<String, Value>::new();
    let env = Environment::new();
    for (key, v) in yaml_context.iter() {
        if let YamlValue::String(key) = key {
            let rendered = env.render_str(v.as_str().unwrap(), &context).unwrap();
            context.insert(key.to_string(), Value::from_safe_string(rendered));
        }
    }
    context
}

fn render_dependencies(
    recipe: &serde_yaml::Mapping,
    context: &BTreeMap<String, Value>,
) -> serde_yaml::Mapping {
    let mut recipe = recipe.clone();

    if let Some(requirements) = recipe.get_mut("requirements") {
        ["build", "host", "run"].iter().for_each(|section| {
            if let Some(YamlValue::Sequence(section)) = requirements.get_mut(section) {
                for item in section {
                    if let YamlValue::String(item) = item {
                        if context.contains_key(item) {
                            let pin = context.get(item).unwrap().as_str().unwrap().to_string();
                            *item = format!("{} {}", item, pin);
                        }
                    }
                }
            }
        });
    }

    recipe
}

/// This renders a recipe, given a variant
/// This evaluates all selectors and jinja statements in the recipe
/// It does _not_ apply the variants to the dependency list yet
pub fn render_recipe(
    recipe: &YamlValue,
    variant: &BTreeMap<String, String>,
) -> anyhow::Result<RenderedRecipe> {
    let recipe = match recipe {
        YamlValue::Mapping(map) => map,
        _ => panic!("Expected a map"),
    };

    let mut env = Environment::new();
    env.add_function("compiler", functions::compiler);
    env.add_function("pin_subpackage", functions::pin_subpackage);
    let (mut recipe_modified, context) =
        if let Some(YamlValue::Mapping(map)) = &recipe.get("context") {
            let mut context = render_context(map);
            let mut recipe_modified = recipe.clone();
            recipe_modified.remove("context");

            // TODO add more appropriate values here
            context.insert("PYTHON".to_string(), "python".into());

            // add in the variant
            for (key, value) in variant {
                context.insert(key.clone(), Value::from_safe_string(value.clone()));
            }
            (recipe_modified, context)
        } else {
            tracing::info!("Did not find context");
            (recipe.clone(), BTreeMap::new())
        };

    render_recipe_recursively(&mut recipe_modified, &env, &context);
    recipe_modified = render_dependencies(&recipe_modified, &context);
    Ok(serde_yaml::from_value(YamlValue::from(recipe_modified))?)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_render_context() {
        let context = r#"
        name: "foo"
        version: "1.0"
        version_name: "{{ name }}-{{ version }}"
        "#;
        let context = serde_yaml::from_str(context).unwrap();
        let context = render_context(&context);
        insta::assert_yaml_snapshot!(context);
    }

    #[test]
    fn test_render() {
        let recipe = r#"
        context:
            name: "foo"
            version: "1.0"
        source:
            - url: "https://example.com/{{ name }}-{{ version }}.tar.gz"
              sha256: "1234567890"
              patches:
            - url: "https://example.com/{{ name }}-{{ version }}.patch"
              sha256: "1234567890"
        package:
            name: "{{ name }}-{{ version }}"
            version: "{{ version }}"
        requirements:
        about:
        "#;
        let recipe = serde_yaml::from_str(recipe).unwrap();
        let recipe = render_recipe(&recipe, &BTreeMap::new());
        assert!(recipe.is_ok());
        insta::assert_yaml_snapshot!(recipe.expect("could not render recipe"));
    }
}
