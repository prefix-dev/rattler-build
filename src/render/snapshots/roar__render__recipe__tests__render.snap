---
source: src/render/recipe.rs
expression: "recipe.expect(\"could not render recipe\")"
---
package:
  name: foo-1.0
  version: "1.0"
source:
  - url: "https://example.com/foo-1.0.tar.gz"
    sha256: "1234567890"
    patches: ~
    folder: ~
  - url: "https://example.com/foo-1.0.patch"
    sha256: "1234567890"
    patches: ~
    folder: ~
build:
  number: 0
  string: h12341234_0
  script: ~
  ignore_run_exports: ~
  ignore_run_exports_from: ~
  run_exports: ~
  noarch: false
  entry_points: []
requirements:
  build: []
  host: []
  run: []
  constrains: []
about:
  home: ~
  license: ~
  license_file: ~
  license_family: ~
  summary: ~
  description: ~
  doc_url: ~
  dev_url: ~
test: ~

